### PR TITLE
WT-8246 Create a csuite test to run compact and RTS

### PR DESCRIPTION
**Pull request by raviprakashgiri29**

### CODE REVIEW

File: test/suite/test_compact04.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+# Test to run compact and rollback to stable operation that performs as expected following a server crash and
+# recovery. Verify that
+#   (a) the on-disk value is replaced by the correct value from the history store, and
+#   (b) newer updates are removed.
+class test_compact04(test_rollback_to_stable_base):
+    session_config = 'isolation=snapshot'
+
+    key_format_values = [
+        ('integer_row', dict(key_format='i')),
+    ]
```
[the remark this answers is not in the document]
Comment (raviprakashgiri29) [reply test/suite/test_compact04.py]: @kommiharibabu I have fixed it in the new csuite test.

File: test/suite/test_compact04.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        self.conn.set_timestamp('oldest_timestamp=' + self.timestamp_str(10) +
+            ',stable_timestamp=' + self.timestamp_str(10))
+
+        value_a = "aaaaa" * 100
+        value_b = "bbbbb" * 100
+        value_c = "ccccc" * 100
+        value_d = "ddddd" * 100
+
+        # Perform several updates.
+        self.large_updates(uri, value_d, ds, nrows, self.prepare, 20)
```
[the remark this answers is not in the document]
Comment (raviprakashgiri29) [reply test/suite/test_compact04.py]: @hbokhari I have fixed in the new csuite test.

File: test/csuite/wt8246_compact_rts_data_correctness/main.c [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+
+    /* Get the stable timestamp from the stable timestamp of the last successful checkpoint. */
+    testutil_check(conn->query_timestamp(conn, ts_string, "get=stable"));
+    testutil_timestamp_parse(ts_string, &stable_ts);
+
+    /* Get the oldest timestamp from the oldest timestamp of the last successful checkpoint. */
+    testutil_check(conn->query_timestamp(conn, ts_string, "get=oldest"));
+    testutil_timestamp_parse(ts_string, &oldest_ts);
+
+    /* Verify data is visible and correct after compact operation was
```
Comment (hbokhari) [inline test/csuite/wt8246_compact_rts_data_correctness/main.c:188]: Comment formatting.

File: test/csuite/wt8246_compact_rts_data_correctness/main.c [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+     * killed and RTS is performed in recovery.
+     */
+    check(session, uri, value_a, 20);
```
Comment (hbokhari) [inline test/csuite/wt8246_compact_rts_data_correctness/main.c:191]: Please add a comment about where these timestamps are coming from.

File: test/csuite/wt8246_compact_rts_data_correctness/main.c [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+
+    for (i = 0; i < NUM_RECORDS; i++) {
+        cursor->set_key(cursor, i + 1);
+        testutil_check(cursor->search(cursor));
+        testutil_check(cursor->get_value(cursor, &val1, &val2, &val3, &str_val));
+
+        val_2_size = strlen(str_val);
+
+        testutil_assert(val_1_size == val_2_size);
+        // printf ("str_val : %s\n", str_val);
```
Comment (hbokhari) [inline test/csuite/wt8246_compact_rts_data_correctness/main.c:289]: `printf` statement is commented out.

**Review by hbokhari — COMMENTED** [review]
My understanding of the current code is that the compact will run only once (unless I missed something very obvious). I believe a more interesting scenario would be where we perform compaction repeatedly, the child process is killed randomly in middle and then recovery is performed.

File: test/csuite/wt8246_compact_rts_data_correctness/main.c [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+static const char working_dir_row[] = "WT_TEST.compact-stress-row";
+static const char working_dir_col[] = "WT_TEST.compact-stress-col";
+static const char compact_file_fmt[] = "%s/compact_started";
+static char value_a[] = "AA";
+static char value_b[] = "BB";
+static char value_c[] = "CC";
+static char value_d[] = "DD";
+
+/* Structures definition. */
+struct thread_data {
```
Comment (AlexeyAnisimovAU) [inline test/csuite/wt8246_compact_rts_data_correctness/main.c:58]: `thread_data` doesn't seem to be used anywhere in the file.

File: test/csuite/wt8246_compact_rts_data_correctness/main.c [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+
+        /*
+         * We do not expect test to reach here. The child process should have been killed by the
+         * parent process.
+         */
+        printf("Child finished processing...\n");
+        return (EXIT_FAILURE);
+    }
+
+    sprintf(ckpt_file, compact_file_fmt, home);
```
Comment (AlexeyAnisimovAU) [inline test/csuite/wt8246_compact_rts_data_correctness/main.c:157]: It's better to use `__wt_snprintf` here. There's another usage in `workload_compact`.

File: test/csuite/wt8246_compact_rts_data_correctness/main.c [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+    while (stat(ckpt_file, &sb) != 0)
```
Comment (AlexeyAnisimovAU) [inline test/csuite/wt8246_compact_rts_data_correctness/main.c:158]: A comment explaining what we're waiting for here would be good.

File: test/csuite/wt8246_compact_rts_data_correctness/main.c [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+        testutil_sleep_wait(1, pid);
+
+    /* Sleep for a while. Let the child process do some operations on the tables. */
+    sleep(TIMEOUT);
```
Comment (AlexeyAnisimovAU) [inline test/csuite/wt8246_compact_rts_data_correctness/main.c:162]: Suggestion: would it be better if we sleep for a random amount of time?

File: test/csuite/wt8246_compact_rts_data_correctness/main.c [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+
+    printf("Removing records...\n");
+    testutil_check(session->open_cursor(session, uri, NULL, NULL, &cursor));
+
+    /* Remove 1/3 of the records from the middle of the key range. */
+    for (i = NUM_RECORDS / 3; i < (NUM_RECORDS * 2) / 2; i++) {
+        testutil_check(session->begin_transaction(session, NULL));
+        cursor->set_key(cursor, i);
+        testutil_check(cursor->remove(cursor));
+        testutil_check(__wt_snprintf(tscfg, sizeof(tscfg), "commit_timestamp=%d", commit_ts));
```
Comment (AlexeyAnisimovAU) [inline test/csuite/wt8246_compact_rts_data_correctness/main.c:369]: Little optimization: we can move the line above outside of the loop.

File: test/csuite/wt8246_compact_rts_data_correctness/main.c [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+    __wt_random_init_seed((WT_SESSION_IMPL *)session, &rnd);
+    testutil_check(session->open_cursor(session, uri, NULL, NULL, &cursor));
+
+    for (i = 0; i < NUM_RECORDS; i++) {
+        testutil_check(session->begin_transaction(session, NULL));
+        cursor->set_key(cursor, i + 1);
+        val = (uint64_t)__wt_random(&rnd);
+        cursor->set_value(cursor, val, val, val, value);
+        testutil_check(cursor->insert(cursor));
+        testutil_check(__wt_snprintf(tscfg, sizeof(tscfg), "commit_timestamp=%d", commit_ts));
```
Comment (AlexeyAnisimovAU) [inline test/csuite/wt8246_compact_rts_data_correctness/main.c:316]: Same here, it's safe to move the timestamp string formatting outside of the loop.

**Review by AlexeyAnisimovAU — COMMENTED** [review]
I have a few questions and suggestions.

File: test/csuite/wt8246_compact_rts_data_correctness/main.c (hunk 7 above)
Comment (raviprakashgiri29) [reply test/csuite/wt8246_compact_rts_data_correctness/main.c]: Fixed

File: test/csuite/wt8246_compact_rts_data_correctness/main.c (hunk 8 above)
Comment (raviprakashgiri29) [reply test/csuite/wt8246_compact_rts_data_correctness/main.c]: Fixed

File: test/csuite/wt8246_compact_rts_data_correctness/main.c (hunk 10 above)
Comment (raviprakashgiri29) [reply test/csuite/wt8246_compact_rts_data_correctness/main.c]: Fixed

File: test/csuite/wt8246_compact_rts_data_correctness/main.c (hunk 11 above)
Comment (raviprakashgiri29) [reply test/csuite/wt8246_compact_rts_data_correctness/main.c]: Fixed

**Comment by raviprakashgiri29** [comment]
> My understanding of the current code is that the compact will run only once (unless I missed something very obvious). I believe a more interesting scenario would be where we perform compaction repeatedly, the child process is killed randomly in middle and then recovery is performed.

Yeah, the compact is started in a child process, interrupted/killed in the middle and then recovery is run on the data files to check the data correctness.

File: test/csuite/wt8246_compact_rts_data_correctness/main.c (hunk 9 above)
Comment (raviprakashgiri29) [reply test/csuite/wt8246_compact_rts_data_correctness/main.c:161]: This is a configured/tested timeout to make sure that the compact runs on the tables and moves pages before it is killed by the parent.

File: test/csuite/wt8246_compact_rts_data_correctness/main.c [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+    return (EXIT_SUCCESS);
+}
+
+static void
+run_compact(WT_SESSION *session, const char *uri)
+{
+    printf("Compact start...\n");
+    /* Perform compact operation. */
+    testutil_check(session->compact(session, uri, NULL));
+    printf("Compact end...\n");
```
Comment (hbokhari) [inline test/csuite/wt8246_compact_rts_data_correctness/main.c:109]: Ideally we want the child process to be killed before this message gets printed. Have you confirmed this?

**Review by AlexeyAnisimovAU — APPROVED** [review]
LGTM!

File: test/csuite/wt8246_compact_rts_data_correctness/main.c (hunk 12 above)
Comment (raviprakashgiri29) [reply test/csuite/wt8246_compact_rts_data_correctness/main.c:109]: @hbokhari Yes, I ran multiple times in my local machine and confirmed that the compact is killed before printing `Compact end...`. It prints `Compact start...` and `Compact process interrupted and killed...`

**Review by hbokhari — APPROVED** [review]
LGTM

File: test/csuite/wt8246_compact_rts_data_correctness/main.c (hunk 12 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt8246_compact_rts_data_correctness/main.c:109]: I guess it depends on the hardware the test is running on and on a general load of the system. Technically there's a possibility that the parent process gets starved to the point where compact has finished, or compact finishes faster than 4 sec. But if that happens we'll see `Child finished processing...` message and the child process will fail with `EXIT_FAILURE`.

File: test/csuite/wt8246_compact_rts_data_correctness/main.c [hunk 13]
```diff
@@ -0,0 +1,378 @@
...
+    large_updates(session, uri, value_c, 40);
+    large_updates(session, uri, value_d, 50);
+
+    /* Verify data is visible and correct. */
+    check(session, uri, value_a, 20);
+    check(session, uri, value_b, 30);
+    check(session, uri, value_c, 40);
+    check(session, uri, value_d, 50);
+
+    /* Pin stable to timestamp 40. */
```
Comment (kommiharibabu) [inline test/csuite/wt8246_compact_rts_data_correctness/main.c:245]: The comment is saying stable timestamp as 40, but the setting is done with 30.

**Review by kommiharibabu — APPROVED** [review]
LGTM with a comment typo fix.

File: test/csuite/wt8246_compact_rts_data_correctness/main.c (hunk 13 above)
Comment (raviprakashgiri29) [reply test/csuite/wt8246_compact_rts_data_correctness/main.c]: Fixed

**Pull request merged**